### PR TITLE
Issue #5037: handle synchronized blocks in FallThroughCheck

### PR DESCRIPTION
**Pull request by cushon**
Issue #5037

### CODE REVIEW

**Comment by Vladlis** [comment]
Please provide regression report on a few projects https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-tester

**Comment by cushon** [comment]
> Please provide regression report 

$ groovy diff.groovy --localGitRepo ~/src/checkstyle --baseBranch master --patchBranch fallthrough-synchronized --config my_check.xml --listOfProjects projects-to-test-on.properties
...
http://cushon.org/reports/diff/index.html

**Comment by rnveach** [comment]
@cushon Please modify `projects-to-test-on.properties` and expand it to more/all projects.
One project isn't usually enough to find issue.

**Comment by rnveach** [comment]
@cushon Also please modify `config my_check.xml` for the check you are testing.
You are not testing `ThrowsCount`.

**Comment by cushon** [comment]
@rnveach thanks, the corrected diff is at: http://cushon.org/reports/diff/index.html

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java [hunk 1]
```diff
@@ -310,6 +313,20 @@ private boolean checkSwitch(final DetailAST literalSwitchAst, boolean useContinu
...
     /**
+     * Checks if a given synchronized block terminated by return, throw or,
+     * if allowed break, continue.
+     * @param synchronizedAst synchronized block to check.
+     * @param useBreak should we consider break as terminator.
+     * @param useContinue should we consider continue as terminator.
+     * @return true if synchronized block is terminated.
+     */
+    private boolean checkSynchronized(final DetailAST synchronizedAst, boolean useBreak,
+        boolean useContinue) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: please indent this line by `(`

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough.java [hunk 2]
```diff
@@ -521,4 +521,14 @@ void tryResource() throws Exception {
...
+
+    void synchronizedStatement() {
+       switch (hashCode()) {
+           case 1:
+               synchronized (this) {
+                   break;
+               }
+           default:
+       }
+    }
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough.java]: one test is not enough, please add test that do cause violations with `synchronized`.
and one more compicated test with `synchronized` nested in `if` for example or smth similar.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java (hunk 1 above)
Comment (cushon) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough.java (hunk 2 above)
Comment (cushon) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough.java]: done

**Comment by romani** [comment]
@rnveach , please do final review.

**Comment by rnveach** [comment]
As CI passes we can merge.

**Comment by rnveach** [comment]
@cushon wercker failure was fixed in master already, please rebase on latest master.

**Comment by cushon** [comment]
@rnveach I rebased it.

**Pull request merged**